### ISSUE TITLE
Dual I2S bus and XMOS support

### ISSUE THREAD

**Issue by tetele** [justLV/onju-voice#74]
### I2S sampling rate

The current hardware uses a shared I2S bus for both microphones and speaker, which dictates that both input and output must share the sampling rate.

There are quite a few people in the Home Assistant community which use this board to create a local [voice assistant](https://www.esphome.io/components/voice_assistant) using ESPHome. I've personally created a few version of ESPHome [configs](https://github.com/tetele/onju-voice-statellite) for it so far.

Most components in ESPHome use a 16kHz sampling rate for audio input - `voice_assistant`, [`micro_wake_word`](https://www.esphome.io/components/micro_wake_word). Which becomes a problem if you want to use the Onju Voice as a media player, as that would constrain it to producing 16kHz output, which sounds.... bad :)

### Echo cancelling

Additionally, micro wake word is something that needs to constantly listen in order to detect wake words, but media playback without echo cancellation would hinder all such efforts, given how there is no way for the microphones not to listen to the audio output constantly.

## Proposal

It would be lovely if we were to have another hardware revision with the following 2 improvements, which would solve the above:
1. separate I2S buses for audio input and output, as there are enough spare ports on the ESP32S3
2. an onboard XMOS chip (or equivalent) which would cancel out the output echo

I have read that you are not willing to work on any improvements, @justLV. However, since fixing the issues above also requires the distribution of such a new hardware revision, I feel like the hardware license (i.e. the NC part of it) prohibits others from tinkering with the board any further, while also making the results available at an accessible price point.

This is not the kind of initiative in which one can contribute and then just let others produce their own PCBs. History has shown that it's hard/costly for people to produce their own PCBs, especially as complex as these.

With that in mind, is there any chance you'll consider either a new hardware revision OR relaxing the hardware license to allow others to manufacture and sell such new revisions that they create? Or let me know if I just don't understand licenses and that's already possible. 😅

**Comment by justLV** [comment]
Thanks for the detailed breakdown!

Yes this actually would be quite neat and fully support. AEC is pretty essential for a lot of applications.

I would only question that the NC part discourages tinkering, is there a specific case of someone in the community who would make this change for the community, but being discouraged from making the revision due to the license if they were planning on open sourcing it anyway? (The task of changing this for someone familiar with the EDA tools is quite minimal tbh)

**Comment by justLV** [comment]
Btw, if someone wants to help do the specification of what chip to use for maximum firmware compatibility across both Arduino and HA, and the proposed schematic changes, I'm happy to look at doing these changes and updating the PCB!

Worth noting though that a dedicated chip like the XVF3800 would have a BOM cost of $15-20 (compared to ESP32 being <$2) fwiw

Also AEC should be possible in software on the ESP32 as is using ESP-ADF, but I understand this incurs extra work

**Comment by denisjoshua** [comment]
We can take ideea from the Home Assistant Voice PE that was presented yesterday... 
It's nice, have a nice hardware but the speakers is very poor... 
So using the Google Nest Mini (a nice and good case with a good speaker) and the hardware/software of Voice PE 
can be a nice Satellite for owner rooms.

**Comment by tetele** [comment]
> I would only question that the NC part discourages tinkering, is there a specific case of someone in the community who would make this change for the community, but being discouraged from making the revision due to the license if they were planning on open sourcing it anyway?

My point was that the ones who'd have to then take care of manufacturing and distribution (mass-manufacturing to lower the production costs) would bump into the license. If you're willing to adopt the new PCB design and put up a new revision on PCBWay, then that point is no longer valid.

> Btw, if someone wants to help do the specification of what chip to use for maximum firmware compatibility across both Arduino and HA, and the proposed schematic changes, I'm happy to look at doing these changes and updating the PCB!

It'll take just a little while longer until HA release the schematics on the Voice PE and we can all learn how they use the XMOS chip. I'll get back then.

**Comment by Hedda** [comment]
I believe that this is more or less a duplicate of this existing feature request for XMOS discussion:

- https://github.com/justLV/onju-voice/issues/59

That too is asking for a new open-source board design that is hardware compatible with the Home Assistant Voice Preview Edition.

**Comment by Hedda** [comment]
FYI; more on the two i2s channel design is also discussed here:

- https://github.com/orgs/FutureProofHomes/discussions/6

**Comment by jhbruhn** [comment]
By now the schematics to the Voice PE have been released, which includes the exact part number of the XMOS chip in use there: https://voice-pe.home-assistant.io/resources/home_assistant_voice_pe_schematic_v1.0_241009.pdf

I would also love to contribute to a design which uses the XMOS chip and two separate I2S channels. My approach would be to take the current board design of Onju Voice, keep the ESP32-S3, its flash, the two microphones (HA Voice PE uses the same ones!) and the LEDs (with 5V supply this time!) and probably also keep the DAC+Amp setup, and then copy over most of the HA Voice PE schematics, mainly the XMOS chip and its flash, with its connections to the ESP32 and the DAC.

**Comment by jhbruhn** [comment]
BTW some additional thoughts about selling ready-made PCBs: While I think there would be a business case here, selling these would require the seller to also get a FCC/... Certification because it is a radio device. Using one of the ESP32 modules would circumvent that, but I don't think they fit on this PCB height-wise.

**Comment by jhbruhn** [comment]
I have started to create a schematic based on the Voice PE. I then incorporated all parts relevant to the Google Home Mini device from the Onju schematic and also prepared the basic board layout. Because I wanted to use KiCad, I created everything from scratch again. Thus I am not even sure whether the original license applies to this anymore.

Before I start routing the PCB it would be nice if someone familiar with ESP32s could sanity-check this schematic. Maybe there is something obvious I forgot or where I made a mistake. I can also provide the schematic files but for now I have this PDF export:

[onju.pdf](https://github.com/user-attachments/files/18266376/onju.pdf)


One thing to keep in mind about the XMOS based design is that it will be more difficult to program than the plain ESP32-S3 design. One needs either the XMOS debug tool or a SPI flash clip to program the initial firmware onto the SPI Flash of the XMOS chip. Only then can the USB or I2C connection be used for DFU, as that is implemented in the firmware rather than a ROM Bootloader like the ESP32 series uses it. Thus, I don't know whether this project would be well suitable with the previously used distribution service via PCBWays PCBA service, unless they could also program the flash chips. I think they have this as a service, but it might be cost-prohibitive.

EDIT: I actually found some problems myself and just updated the schematic.

**Comment by alextrical** [comment]
I'm happy to take a look over your schematic and give some assistance/pointers, as your effectively going down a very similar route as I was many months ago, with LibreTalk Wave / Signal

@jhbryhn If your on Discord it would be great to see you on the FutureProofHomes server :)

**Comment by jhbruhn** [comment]
Thanks to your comment in the other issue, the Flashing issue for the XMOS chip should be solved with the Mux between ESP32s SPI Bus and and the XMOS Flash Chip. See the attached schematic here for my implementation of it. It's a bit different from yours as I won't be using the SPI Bus to interface with the XMOS controller, so it only Muxes between the XU316 Flash Connection and the ESP32 SPI Bus. But similar results should be achievable.

If you have some time to spare I'd be glad if you could have a look at the updated schematic:

[onju.pdf](https://github.com/user-attachments/files/18269776/onju.pdf)

**Comment by Hedda** [comment]
> Because I wanted to use KiCad, I created everything from scratch again. Thus I am not even sure whether the original license applies to this anymore.

Awesome that you choose to use KiCad as that will likely mean it will be more accessible to other people can assist.

If acceptable and you have no commercial aspirations then suggest that you use some permissive open source license like example the same MIT license that @justLV used for this onju-voice project (which would allow anyone else to sell boards with or without modifications).

* https://www.oshwa.org/faq/

> I don't know whether this project would be well suitable with the previously used distribution service via PCBWays PCBA service, unless they could also program the flash chips. I think they have this as a service, but it might be cost-prohibitive.

While not a must, I believe that that it would be ideal to have a foolproof simple design that anyone could order from PCBWay, JLCPCB, OSH Park, OSHWLab, and similar PCB services that have "shared projects" where they always have they parts in stock and manufacturering is simple enough to have low failure rate.

* https://hackaday.com/2023/05/09/share-your-projects-making-helpful-pcbs/

  * https://www.pcbway.com/project/shareproject/Onju_Voice_d33625a1.html

> _BTW some additional thoughts about selling ready-made PCBs: While I think there would be a business case here, selling these would require the seller to also get a FCC/... Certification because it is a radio device. **Using one of the ESP32 modules would circumvent that, but I don't think they fit on this PCB height-wise.**_

Hmm, do you think the hight within a Google Home Mini would really be a real design problem using ESP32 modules? I though many ESP32 radio modules were very low profile. 

To work around that specific issue would it be possible to split away the ESP32 module part as a separate PCB and some how connect it via a side-connector or via a flat ribbon-cable or other similar solutions to mount it side-by-side without adding any hight?

Alternative could it be possible to fit an ESP32 modules if used a "Cavity PCB" design (also known as "recessed PCB" design) and put the ESP32 inside that cavity? Cavity PCB design which as you probably already know refers to cutting one or more cavities or recesses within the board itself?

Also remember seeing someone making a cut-out design (which was for a Raspberry Pi Compute Module carrier board) where they placed the module inside the cut-out section and connecting it via some kind PCB that bridged the gap and had connectors in order to make the whole package be low profile.

**Comment by jhbruhn** [comment]
As explained in this issue already, the hardware of this project is licensed with a different, non-commercial license.

About the modules: I will have a look at it. But I currently do not plan on selling these boards.

**Comment by Hedda** [comment]
> As explained in this issue already, the hardware of this project is licensed with a different, non-commercial license.

Different? non-commercial license? Is it not just using the standard MIT License? That is permissive and allows commercial use. The only real condition that the standard MIT License has is requiring preservation of any copyright and license notices.

* https://github.com/justLV/onju-voice/blob/master/LICENSE

As explained in FAQ from the OSHW (Open Source Hardware Association), non-commercial restrictions are not open source HW:

* https://www.oshwa.org/faq/#non-commercial

Using a non-commercial is a bad idea and then it is not possbile for someone to release design based on this and your project.

**Comment by jhbruhn** [comment]
Here the hardware is licensed under CC-BY-NC: https://www.pcbway.com/project/shareproject/Onju_Voice_d33625a1.html

But this is in fact interesting. In the repository the License is MIT and the PCB files should also be licensed MIT.

On PCBWay only the Gerber and BOM files are licensed with CC-BY-NC.

My derivative is only based on the design files in the repository, if at all, which technically are licensed with MIT. So technically I could also license it under MIT or perhaps a proper hardware license.

But to act fairly maybe justLV should tell what his intentions with the licensing were.

**Comment by tetele** [comment]
> But this is in fact interesting. In the repository the License is MIT and the PCB files should also be licensed MIT.

The software in the repo is MIT, not the hardware design :D

**Comment by jhbruhn** [comment]
> The software in the repo is MIT, not the hardware design :D

I also believe that was justLVs intention, but can you give me a reference to that? Because in this GitHub Repo, there is just the MIT License, under which thus the hardware design files that are part of this repository are published.

Only the uploaded files on PCBWay are licensed with CC-BY-NC.

**Comment by tetele** [comment]
![20250103_131249](https://github.com/user-attachments/assets/ec108a90-cb04-49d3-8c36-e9614816df2e)

**Comment by jhbruhn** [comment]
...makes sense. And I had the PCB lying in front of me on my desk all the time :D 🤦

**Comment by Hedda** [comment]
> Because in this GitHub Repo, there is just the MIT License, under which thus the hardware design files that are part of this repository are published.
> 
> Only the uploaded files on PCBWay are licensed with CC-BY-NC.

@justLV This fact that the repo here on GitHub only has an MIT license file (which can also be used for hardware) does make it really confusing to understand what license applies to fact, so you should really add additional license files to each directory/folder to clearify this better.

Updated comment: Sad to find out that the hardware is not fully open source as that is again what FAQ from the OSHW (Open Source Hardware Association) recommend as in their opinion non-commercial restrictions should not count as open source HW -> https://www.oshwa.org/faq/#non-commercial

### *Why aren’t non-commercial restrictions compatible with open source hardware?*

*There are a few reasons.*

*If you place a non-commercial restriction on your hardware design, other people don’t have the same freedom to use the design in the ways that you can. That means, for example, that if you and someone else both release designs with non-commercial licenses, neither of you can make and sell hardware that builds on both of your designs. Rather than contributing to a commons of hardware designs for everyone to build on, you’re limiting others to a very narrow range of possible uses for your design.*

*In particular, because making hardware invariably involves money, it’s very difficult to make use of a hardware design without involving some commercial activity. For example, say a group of friends wanted to get together and order ten copies of a hardware design – something that’s often much cheaper than each person ordering their own copy. If one person places the order and the others pay him back for their share, they’d probably be violating a non-commercial restriction. Or say someone wants to charge people to take a workshop in which they make and keep a copy of your hardware design – that’s also commercial activity. In general, there are just very few ways for someone to use a hardware design without involving some sort of commercial activity.*

**Comment by alextrical** [comment]
Its one of the issues we are having to deal with over at FutureProofHomes with the Satellite1 too. Essentially wanting to make the design reviewable and accessible to individuals, but not wanting clones to be created and undercutting the sales that fund further development of the design. Licensing is a PITA. Essentially we will be releasing the schematics as open source, and the kicad/gerber files on a delay (Likely 12 months) to help mitigate the issues with clones.

**Comment by Hedda** [comment]
Yeah, but if you use a license that restrict usage then not call it ”open source hardware”, and be aware that you using non-commercial license on hardware that you refer to as ”open source hardware” can be a disservice to others as they could be acused of violating your license even if there are only so many ways that you can design these type of devices.

**Comment by nanosonde** [comment]
I really really did not want to comment on this, but sorry I have to before exploding internally....

I think that designs like this are too often too overestimated with respect to complexity and the related creativity level that the creator/dev wants to "protect" now.
Each of our electronic engineers could create such a design in probably less than a week. There is really no rocket science involved (see eval designs, app notes, etc). Stable working software here is the complex part.

So I think that it is best to accept that there will be clones. Presumably they will come from Asia and from our experience there is little one can do about it. 
You have already published photos of the PCB, I would not be astonished if there are people out there who already create their own designs based on this.

Please do not get me wrong: I really appreciate your efforts and I will probably order some satellite once it becomes available in the EU. But please make it a real open hardware design. It makes me feel better to support such projects and even if there would be cheaper clones, I would not buy them! ;-)

UPDATE: Sorry, was confused and mentioned the FPH/satellite here, but the same applies to this HW (onju-voice) here. :-)

**Comment by jhbruhn** [comment]
I don't want to kill this discussion, but I think it might be better suited for a different context. This issue is partially about the license of the Onju Voice project, not about the Satellite 1.

Also don't underestimate that Hardware and Software Design is only one part (and a rather small part in a lot of cases) of making a product.

**Comment by TechnoTaff** [comment]
Maybe a new project, more of a fork of Voice-PR in Google Home Mini form-factor - rather than a new version of Onju would be more suitable.

If it was principally the actual Voice-PE hardware design (updated to support the 4-LED output, touch input and circular form-factor) then it might be easier to consume future Voice-PE community improvements?

**Comment by jhbruhn** [comment]
That is basically what the design I am working on resembles.

Right now I am working on getting the manufacturing cost down, aiming for a similar model like Onju does it by having the project as a PCBWay Project that can easily be ordered on demand.

I have not published the design yet as I first want to test it properly, and check the cost-feasibility. If that fails (i.e. cost in single quantities is higher than just getting a Voice PE, +-), I will probably publish it anyways but won't pursue it further.

**Comment by bluenazgul** [comment]
Its maybe a stupid idea, but would it be possible to modify the "Onju Voice" to have the same Base Components as the "Voice PE"? Since the shematic and components of the Voice PE are known

**Comment by jhbruhn** [comment]
As I've already written a couple of times, that is basically what I am doing right now.

**Comment by TechnoTaff** [comment]
> Its maybe a stupid idea, but would it be possible to modify the "Onju Voice" to have the same Base Components as the "Voice PE"? Since the shematic and components of the Voice PE are known

I think you're describing the same thing. But at what point is it still Onju Voice? Is Onju Voice the voice hardware or the form-factor?

If it's the voice hardware is a clone of Voice-PE, and the form-factor is Google's, is it still Onju Voice?

I think we would have to acknowledge the huge contribution Justin has made on how to interface with Google's form-factor and daughter PCBs, but it doesn't need to be part of the Onju project.

**Comment by bluenazgul** [comment]
I would say it still is, as it still is an Hardware Conversation

**Comment by Hedda** [comment]
> I think you're describing the same thing. But at what point is it still Onju Voice? Is Onju Voice the voice hardware or the form-factor?
> 
> If it's the voice hardware is a clone of Voice-PE, and the form-factor is Google's, is it still Onju Voice?
> 
> I think we would have to acknowledge the huge contribution Justin has made on how to interface with Google's form-factor and daughter PCBs, but it doesn't need to be part of the Onju project.

I agree that this specific issue is not the right place to discuss that (as this issue specific issue is still a hardware discussion) but yes, all you guys wrote is correct. Update: Started new issue for that here to seperate that discussion -> https://github.com/justLV/onju-voice/issues/77

Anyway, no matter who makes it, any new PCB design that is based on the newly released [Home Assistant Voice Preview Edition](https://voice-pe.home-assistant.io/resources/) circuit diagram/schematics instead of straight-out strictly copying justLV's onju-voice PCB design and circuit diagram can be released under the same fully open-source hardware license as Home Assistant Voice Preview Edition without adding the same no-commercial restriction clause. 

So if [jhbruhn](https://github.com/jhbruhn) or whoever else uses the [Home Assistant Voice PE](https://voice-pe.home-assistant.io/resources/) circuit diagram/schematics as a base but then changes the LED lights and add connection for the mute switch and capacitive touch panels for input, etc. as well as make it a circular side to have it fit as a drop-in replacement in the Google Home Minis / Google Nest Mini then it should not violate the restrictive non-commercial hardware license clause currently set on the onju-voice PCB.

* https://voice-pe.home-assistant.io/resources/

As per Paulus Schoutsen's Home Assistant blog post title and content; "_**The era of open voice assistants has arrived**_"

* https://www.home-assistant.io/blog/2024/12/19/voice-preview-edition-the-era-of-open-voice/

  * https://www.home-assistant.io/blog/2024/12/19/voice-preview-edition-the-era-of-open-voice/#fully-open-and-customizable

    * https://www.home-assistant.io/blog/2024/12/19/voice-preview-edition-the-era-of-open-voice/#community-driven

      * https://www.home-assistant.io/blog/2024/12/19/voice-preview-edition-the-era-of-open-voice/#conclusion

"_We’re not just launching a new product, **we’re open sourcing all of it**. We built this for the Home Assistant community. Our community doesn’t want a single voice assistant, they want the one that works for them – they want choice. **Creating a voice assistant is hard, and until now, parts of the solution were locked behind expensive licenses and proprietary software. With Voice Preview Edition being open source, we hope to bootstrap an ecosystem of voice assistants**._"

**Comment by elsi06** [comment]
Any Update here?

**Comment by jhbruhn** [comment]
No

**Comment by Hedda** [comment]
While not a must it would be very nice and easier for others if @justLV could simply change the HW license of onju-voice hardware PCB design to a non-restrictive fully open-source hardware license without no-commercial restrictions for OSHW compatibility:

* https://github.com/justLV/onju-voice/issues/77

Even if there are no updates on the hardware-side there are at least loads happening on the software-side which could incentivize(?).

- https://www.home-assistant.io/blog/2025/02/13/voice-chapter-9-speech-to-phrase/

- https://github.com/esphome/home-assistant-voice-pe

Recommend to check out this new video for Home Assistant Voice Chapter 9 video that Nabu Casa devs live-streamed yesterday:

* https://www.youtube.com/watch?v=k6VvzDSI8RU

FYI, new speaker media player component with several other helper components has been backported and merged into ESPHome:

- https://github.com/esphome/esphome/commits?author=kahrendt&since=2025-01-01&until=2025-02-13

So much of Nabu Casa's i2s + media player code is upstreamed so will be in the upcoming ESPHome 2025.2 release which is in beta:

- https://github.com/esphome/esphome/releases/

Tip is to set a reminder for your YouTube account to watch the ESPHome live stream next week:

* https://www.youtube.com/live/HtG3fjAO96c

Note that these also depend on the new esp-audio-libs library available via PlatformIO:

 - https://github.com/esphome/esp-audio-libs

PS: Btw, check out this not yet merged pull request which sounds relevant (pun intended) as adds basic support for a new i2s driver:

- https://github.com/esphome/esphome/pull/8181

**Comment by Hedda** [comment]
FYI; iMike78 has begun a new project that will work on open-source PCB designs for Google Nest / Google Home that is using the Home Assistant Voice Preview Edition with XMOS chip as its reference (and with the aim to be hardware compatibile with it), see:

* https://github.com/iMike78/nest-mini-drop-in-pcb

There is currently more ongoing discussion among end-users about that effort in this thread on Home Assistant's community forum;

* https://community.home-assistant.io/t/any-news-on-alternative-to-onju-voice-pcb-repacement-design-for-google-nest-home-mini-speakers-with-added-xmos-chip-to-match-official-home-assistant-voice-preview-edition-reference-hardware/860001/

Note that iMike78 is only working on this in his spare time so he posted there that he really use some help from someone with PCB design experience in GND pouring, shielding, and overall routing, (using KiCad for both the schematic and PCB layout).

**Comment by iMike78** [comment]
Thanks for sharing my work here, @Hedda.
I started working on this about a month ago, but as you mentioned, I can only do it in my spare time.
The schematics and placement are based on @justLV’s Onju Voice project and the Home Assistant Voice PE reference.
It uses dual I2S buses and an XMOS chip for noise cancellation.

@tetele – I might need your contribution later for the software implementation. You've done some great iterations of both the Onju Voice and HA VPE firmware.

**Comment by tetele** [comment]
@iMike78 let me know when/how i can help.

**Comment by iMike78** [comment]
> [@iMike78](https://github.com/iMike78) let me know when/how i can help.

Thanks, @tetele!
I’ll need to finish the hardware first, and then I’ll definitely need your help with the ESPHome part.
The ESP32-S3 and XMOS pinout is mostly the same as in the HA VPE, with a few changes — we don’t have the rotary encoder, but we do have touch controls instead, and only 6 LEDs.

At the moment, we have only two microphones, and the XMOS chip is used only for noise cancellation. That means we can stick with the XMOS firmware from the HA VPE project for now.

In the future, I might add a third mic (there’s physical space for it), and we could offload some control logic to the XMOS chip — like LEDs, buttons, or even audio output — similar to what the Satellite 1 project does. That would help take some load off the ESP32.
Of course, that’s a whole new level — and honestly, I have no idea how to write code for the XMOS. 😅

I’ll let you know when I’m ready — and I’ll try to send you a test board as well for development.

**Comment by Hedda** [comment]
> @tetele – I might need your contribution later for the software implementation. You've done some great iterations of both the Onju Voice and HA VPE firmware.

@iMike78 I am sure there are many others who would be willing to help optimize the software-side once it is possible to order pre-assembled PCB boards directly from all-in-one manufacturer(s) that have some kind of "shared projects" like example PCBWay and OSHWLab which allow you to simply select a pre-uploaded design someone else has already have made and quality tested.

@formatBCE is for example someone else who thinkers a lot with ESPHome voice assistant configuration and custom integrations for the [Seeed Respeaker Lite Voice Kit ](https://community.home-assistant.io/t/respeaker-lite-new-seeed-studio-voice-assistant-development-kit-hardware-combine-esp32-with-xmos-xu316-dsp-chip-for-advanced-audio-processing-as-a-esphome-based-home-assistant-assist-satellite-voice-devkit/756944/)(which also has an XMOS chip), so perhaps he might also be interesting in getting his hands on a prototype board as well once the hardware is ready to be tested. Just check out his [Koala-Satellite](https://github.com/formatBCE/Koala-Satellite) and [Respeaker-Lite-ESPHome-integration](https://github.com/formatBCE/Respeaker-Lite-ESPHome-integration) projects for reference.
